File: sdk/ruby/lib/hitechcloud/http_client.rb
# frozen_string_literal: true

require 'net/http'
require 'json'
require 'uri'

module HiTechCloud
  class HTTPClient
    attr_accessor :token

    def initialize(base_url, timeout: 30, max_retries: 3)
      @base_url = base_url
      @timeout = timeout
      @max_retries = max_retries
      @token = nil
    end

    def request(method, path, body: nil, params: nil)
      last_error = nil

      (0..@max_retries).each do |attempt|
        begin
          url = "#{@base_url}#{path}"
          url += "?#{URI.encode_www_form(params)}" if params && !params.empty?

          uri = URI(url)
          http = Net::HTTP.new(uri.host, uri.port)
          http.use_ssl = uri.scheme == 'https'
          http.read_timeout = @timeout

          case method.to_s.upcase
          when 'GET'
            req = Net::HTTP::Get.new(uri)
          when 'POST'
            req = Net::HTTP::Post.new(uri)
            req.body = body.to_json if body
          when 'PUT'
            req = Net::HTTP::Put.new(uri)
            req.body = body.to_json if body
          when 'DELETE'
            req = Net::HTTP::Delete.new(uri)
          end

          req['Content-Type'] = 'application/json'
          req['User-Agent'] = "HiTechCloud-SDK-Ruby/#{VERSION}"
          req['Authorization'] = "Bearer #{@token}" if @token

          response = http.request(req)

          if response.code.to_i == 429 && attempt < @max_retries
            retry_after = response['Retry-After']&.to_i || 60
            sleep(retry_after)
            next
          end

          if response.code.to_i >= 500 && attempt < @max_retries
            sleep(2**attempt)
            next
          end

          result = JSON.parse(response.body)

          if response.code.to_i >= 400
            HiTechCloud.map_error(response.code.to_i, result)
          end

          return result
        rescue StandardError => e
          last_error = e
          raise if e.is_a?(HiTechCloudError)
          sleep(2**attempt) if attempt < @max_retries
        end
      end

      raise last_error
    end

    def get(path, params: nil)
      request('GET', path, params: params)
    end

    def post(path, body: nil)
      request('POST', path, body: body)
    end

    def put(path, body: nil)
      request('PUT', path, body: body)
    end

    def delete(path)
      request('DELETE', path)
    end
  end
end

File: sdk/python/hitechcloud/http_client.py
"""HiTechCloud SDK HTTP Client"""

import json
import time
from typing import Any, Dict, Optional, Union
from urllib.parse import urljoin

import requests

from .exceptions import HiTechCloudError, RateLimitError, ERROR_MAP


class HTTPClient:
    """Low-level HTTP client for HiTechCloud API"""

    def __init__(
        self,
        base_url: str = "https://api.hitechcloud.vn",
        timeout: int = 30,
        max_retries: int = 3,
    ):
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.max_retries = max_retries
        self.session = requests.Session()
        self.session.headers.update({
            "Content-Type": "application/json",
            "Accept": "application/json",
            "User-Agent": "HiTechCloud-Python-SDK/1.0.0",
        })

    def set_token(self, token: str):
        """Set authorization token"""
        self.session.headers["Authorization"] = f"Bearer {token}"

    def clear_token(self):
        """Remove authorization token"""
        self.session.headers.pop("Authorization", None)

    def request(
        self,
        method: str,
        path: str,
        params: Dict = None,
        data: Dict = None,
        json_data: Dict = None,
        headers: Dict = None,
    ) -> Any:
        """Make HTTP request with retry logic"""
        url = urljoin(self.base_url + "/", path.lstrip("/"))
        last_error = None

        for attempt in range(self.max_retries):
            try:
                response = self.session.request(
                    method=method,
                    url=url,
                    params=params,
                    data=data,
                    json=json_data,
                    headers=headers,
                    timeout=self.timeout,
                )

                if response.status_code == 429:
                    retry_after = int(response.headers.get("Retry-After", 60))
                    if attempt < self.max_retries - 1:
                        time.sleep(retry_after)
                        continue
                    raise RateLimitError(
                        "Rate limit exceeded",
                        status_code=429,
                        retry_after=retry_after,
                    )

                if response.status_code >= 400:
                    self._raise_error(response)

                if response.status_code == 204:
                    return None

                return response.json()

            except requests.exceptions.ConnectionError as e:
                last_error = e
                if attempt < self.max_retries - 1:
                    time.sleep(2 ** attempt)
                    continue
                raise HiTechCloudError(f"Connection failed: {e}")

            except requests.exceptions.Timeout as e:
                last_error = e
                if attempt < self.max_retries - 1:
                    time.sleep(2 ** attempt)
                    continue
                raise HiTechCloudError(f"Request timed out: {e}")

        if last_error:
            raise HiTechCloudError(f"Request failed after {self.max_retries} retries: {last_error}")

    def _raise_error(self, response: requests.Response):
        """Parse error response and raise appropriate exception"""
        try:
            body = response.json()
            errors = body.get("error", [])
            if isinstance(errors, str):
                errors = [errors]
            message = "; ".join(errors) if errors else f"HTTP {response.status_code}"
        except (json.JSONDecodeError, ValueError):
            message = f"HTTP {response.status_code}: {response.text[:200]}"
            errors = [message]

        error_class = ERROR_MAP.get(response.status_code, HiTechCloudError)
        kwargs = {
            "message": message,
            "status_code": response.status_code,
            "errors": errors,
        }

        if response.status_code == 429:
            kwargs["retry_after"] = int(response.headers.get("Retry-After", 60))

        raise error_class(**kwargs)

    def get(self, path: str, params: Dict = None) -> Any:
        return self.request("GET", path, params=params)

    def post(self, path: str, data: Dict = None) -> Any:
        return self.request("POST", path, json_data=data)

    def put(self, path: str, data: Dict = None) -> Any:
        return self.request("PUT", path, json_data=data)

    def delete(self, path: str) -> Any:
        return self.request("DELETE", path)

    def close(self):
        """Close the HTTP session"""
        self.session.close()
